syntax = "proto2";

package hourai.db.proto;

import "hourai/db/proto/auto_config.proto";
import "hourai/db/proto/escalation.proto";
import "hourai/db/proto/util.proto";

message GuildConfig {
  optional LoggingConfig logging = 1;
  optional ValidationConfig validation = 2;
  optional AutoConfig auto = 3;
  optional ModerationConfig moderation = 4;
  optional MusicConfig music = 5;
  optional AnnouncementConfig announce = 6;
  optional RoleConfig role = 7;
}

// ------------------------------------------------------------------------------
// Logging Configs
// ------------------------------------------------------------------------------

message LoggingConfig {
  // Required. The channel to make modlog messages to.
  optional uint64 modlog_channel_id = 1;

  // Optional: If set to true, deleted messages are logged to the modlog channel.
  optional bool log_deleted_messages = 2;
  // Optional: If set to true, edited messages are logged to the modlog channel.
  // optional bool log_edited_messages = 2;
}

// ------------------------------------------------------------------------------
// Moderation Configs
// ------------------------------------------------------------------------------

message ModerationConfig {
  optional EscalationLadder escalation_ladder = 1;
}

// ------------------------------------------------------------------------------
// Music Configs
// ------------------------------------------------------------------------------

message MusicConfig {
  // The volume the music bot operates at.
  optional uint32 volume = 1 [default = 100];
  // Optional: If set, people who can DJ the bot need to have this role.
  // Otherwise, moderators by default have the permission.
  repeated uint64 dj_role_id = 2 [packed = true];
  // Optional: If set, the music bot wil only join that voice channel.
  // Otherwise, the bot will join the first caller's current voice channel.
  optional uint64 voice_channel_id = 3;
  // Optional: If set, music commands will only work in the specified channels.
  repeated uint64 text_channel_id = 4 [packed = true];
}

// ------------------------------------------------------------------------------
// Music Configs
// ------------------------------------------------------------------------------

message AnnouncementConfig {
  optional AnnouncementTypeConfig joins = 1;
  optional AnnouncementTypeConfig leaves = 2;
  optional AnnouncementTypeConfig bans = 3;
  optional AnnouncementTypeConfig streams = 4;
  optional AnnouncementTypeConfig voice = 5;
}

message AnnouncementTypeConfig {
  // Required. The channels to output the announcements to.
  repeated uint64 channel_ids = 1 [packed = true];
  // Optional: Custom messages used. If none are provided, the default message
  // will be used.
  repeated string messages = 2;
}

// ------------------------------------------------------------------------------
// Validation Configs
// ------------------------------------------------------------------------------

message ValidationConfig {
  // If set to true, validation as a whole is enabled.
  optional bool enabled = 1;

  // Optional: The role to use for validation. If not set, no role will be given.
  optional uint64 role_id = 2;

  // Optional: If set, unvalildated users are kicked after the number of seconds
  // passes. Minimum value is 1 hour (3600 seconds).
  optional uint64 kick_unvalidated_users_after = 3;

  // Optional: If set, accounts under the age, in seconds, will be rejected.
  optional uint64 minimum_account_age = 4 [default = 2592000];

  // Optional: aspect specific configurations. If not set, the default values are
  // used.
  optional AvatarValidationConfig avatar = 5;
  optional UsernameValidationConfig username = 6;
  optional CrossGuildValidationConfig cross_server = 7;
}

message AvatarValidationConfig {
  optional bool reject_default_avatars = 1 [default = true];
  // TODO(james7132): Enable this when adding further avatar validation.
  //optional bool reject_sexual_avatars = 1 [default = true];
  //optional bool reject_violent_avatars = 1 [default = true];
}

message UsernameValidationConfig {
  // Optional: The blacklist applies first. If any matches, the user is rejected,
  // unless the any entry in the whitelist applies. Both are lists of regexes.
  optional FilterSettings username_filter = 1;

  optional bool reject_sexual_usernames = 2 [default = true];
  optional bool reject_offensive_usernames = 3 [default = true];
  optional bool reject_likely_user_bots = 4 [default = true];
}

message CrossGuildValidationConfig {
  // Optional: If set to true, users banned from other servers
  optional bool reject_banned_users = 1 [default = true];
  // Optional: This marks the minimum size of other servers to check against. If
  // not set, the default of 150 users is used. This
  optional uint64 minimum_guild_size = 2 [default = 150];
  // TODO(james7132): Update this when the Discord Hotline integration is
  // available.
  // optional bool reject_hotline_reported_users = 3 [default = true];
}

// ------------------------------------------------------------------------------
// Role Configs
// ------------------------------------------------------------------------------
message RoleConfig {
  // Optional: The IDs of the roles that normal users can request whenever.
  repeated uint64 self_serve_role_ids = 1 [packed = true];
}
